File: haskell/day_2/app/Main.hs
module Main where

import Data.List (tails)
import Data.Maybe (mapMaybe)
import GHC.Num (Num)
import System.Environment (getArgs)
import Text.Read (readMaybe)

data Command
  = Forward Integer
  | Up Integer
  | Down Integer
  deriving (Read, Show)

main :: IO ()
main = do
  args <- getArgs
  contents <- readFile $ head args

  let input = mapMaybe parseCommand $ lines contents

  putStrLn $ "Part 1 solution: " ++ show (partOne input)
  putStrLn $ "Part 2 solution: " ++ show (partTwo input)

  return ()

parseCommand :: String -> Maybe Command
parseCommand s = case words s of
  [dir, num] ->
    parseDir dir <*> (readMaybe num :: Maybe Integer)
  _ -> Nothing
  where
    parseDir dir =
      case dir of
        "forward" -> Just Forward
        "up" -> Just Up
        "down" -> Just Down
        _ -> Nothing

partOne :: [Command] -> Integer
partOne commands =
  let (position, depth) = foldl f (0, 0) commands
   in position * depth
  where
    f (position, depth) (Forward c) = (position + c, depth)
    f (position, depth) (Up c) = (position, depth - c)
    f (position, depth) (Down c) = (position, depth + c)

partTwo :: [Command] -> Integer
partTwo commands =
  let (position, depth, _) = foldl f (0, 0, 0) commands
   in position * depth
  where
    f (position, depth, aim) (Forward c) = (position + c, depth + (aim * c), aim)
    f (position, depth, aim) (Up c) = (position, depth, aim - c)
    f (position, depth, aim) (Down c) = (position, depth, aim + c)

File: rust/day_2/src/main.rs
use std::env;
use std::fmt::{Display, Formatter};
use std::fs;
use std::str::FromStr;

#[derive(Debug)]
struct ParseError;

impl Display for ParseError {
    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
        write!(f, "invalid movement input")
    }
}

impl From<std::num::ParseIntError> for ParseError {
    fn from(_: std::num::ParseIntError) -> Self {
        ParseError
    }
}

enum Movement {
    Forward(u32),
    Up(u32),
    Down(u32),
}

impl FromStr for Movement {
    type Err = ParseError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s
            .split(" ")
            .map(|s| s.trim())
            .collect::<Vec<&str>>()
            .as_slice()
        {
            [dir, count] => {
                let count = count.parse::<u32>()?;

                match *dir {
                    "forward" => Ok(Self::Forward(count)),
                    "up" => Ok(Self::Up(count)),
                    "down" => Ok(Self::Down(count)),
                    _ => Err(ParseError),
                }
            }

            _ => Err(ParseError),
        }
    }
}

fn main() {
    let mut args: Vec<String> = env::args().collect();

    // Trash command args
    args.remove(0);

    if let Some(filename) = args.pop() {
        let contents = fs::read_to_string(filename).expect("unable to read file");
        let movements: Vec<Movement> = contents
            .lines()
            .map(|l| l.parse::<Movement>())
            .collect::<Result<Vec<Movement>, ParseError>>()
            .expect("failed to parse movements");

        println!("Part 1 answer: {}", solve_part_1(&movements));
        println!("Part 2 answer: {}", solve_part_2(&movements));
    }
}

fn solve_part_1(movements: &Vec<Movement>) -> i32 {
    let (position, depth): (i32, i32) =
        movements
            .iter()
            .fold((0, 0), |(position, depth), movement| match *movement {
                Movement::Forward(count) => (position + count as i32, depth),
                Movement::Up(count) => (position, depth - count as i32),
                Movement::Down(count) => (position, depth + count as i32),
            });

    position * depth
}

fn solve_part_2(movements: &Vec<Movement>) -> i32 {
    let (position, depth, _aim): (i32, i32, i32) = movements.iter().fold(
        (0, 0, 0),
        |(position, depth, aim), movement| match *movement {
            Movement::Forward(count) => {
                (position + count as i32, depth + (aim * count as i32), aim)
            }

            Movement::Up(count) => (position, depth, aim - count as i32),
            Movement::Down(count) => (position, depth, aim + count as i32),
        },
    );

    position * depth
}
